# Student Performance Prediction using OULAD Dataset

### Research Preprocessing Notebook

**Student:** Eman Ali

**Dataset:** Open University Learning Analytics Dataset (OULAD)

# Step 1: Import Libraries

In [1]:
import pandas as pd
import numpy as np
import os

# Step 2: Load Dataset

In [2]:
courses = pd.read_csv("courses.csv")
assessments = pd.read_csv("assessments.csv")
studentAssessment = pd.read_csv("studentAssessment.csv")
studentInfo = pd.read_csv("studentInfo.csv")
studentRegistration = pd.read_csv("studentRegistration.csv")
studentVle = pd.read_csv("studentVle.csv")
vle = pd.read_csv("vle.csv")

# Step 3: Explore Dataset

In [3]:
print("courses:", courses.shape)
print("assessments:", assessments.shape)
print("studentAssessment:", studentAssessment.shape)
print("studentInfo:", studentInfo.shape)
print("studentRegistration:", studentRegistration.shape)
print("studentVle:", studentVle.shape)
print("vle:", vle.shape)

courses: (22, 3)
assessments: (206, 6)
studentAssessment: (173912, 5)
studentInfo: (32593, 12)
studentRegistration: (32593, 5)
studentVle: (10655280, 6)
vle: (6364, 6)


# Step 4: Explore studentInfo Dataset

In [4]:
studentInfo.info()

<class 'pandas.DataFrame'>
RangeIndex: 32593 entries, 0 to 32592
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype
---  ------                --------------  -----
 0   code_module           32593 non-null  str  
 1   code_presentation     32593 non-null  str  
 2   id_student            32593 non-null  int64
 3   gender                32593 non-null  str  
 4   region                32593 non-null  str  
 5   highest_education     32593 non-null  str  
 6   imd_band              31482 non-null  str  
 7   age_band              32593 non-null  str  
 8   num_of_prev_attempts  32593 non-null  int64
 9   studied_credits       32593 non-null  int64
 10  disability            32593 non-null  str  
 11  final_result          32593 non-null  str  
dtypes: int64(3), str(9)
memory usage: 3.0 MB


In [5]:
studentInfo.shape

(32593, 12)

In [6]:
studentInfo.columns

Index(['code_module', 'code_presentation', 'id_student', 'gender', 'region',
       'highest_education', 'imd_band', 'age_band', 'num_of_prev_attempts',
       'studied_credits', 'disability', 'final_result'],
      dtype='str')

In [7]:
studentInfo.isnull().sum()

code_module                0
code_presentation          0
id_student                 0
gender                     0
region                     0
highest_education          0
imd_band                1111
age_band                   0
num_of_prev_attempts       0
studied_credits            0
disability                 0
final_result               0
dtype: int64

In [8]:
studentInfo["final_result"].value_counts()

final_result
Pass           12361
Withdrawn      10156
Fail            7052
Distinction     3024
Name: count, dtype: int64

# Step 5: Clean studentInfo Dataset

In [9]:
studentInfo.duplicated().sum()

np.int64(0)

In [10]:
studentInfo = studentInfo.drop_duplicates()
studentInfo.shape

(32593, 12)

In [11]:
studentInfo["imd_band"] = studentInfo["imd_band"].fillna("Unknown")

In [12]:
studentInfo.isnull().sum()

code_module             0
code_presentation       0
id_student              0
gender                  0
region                  0
highest_education       0
imd_band                0
age_band                0
num_of_prev_attempts    0
studied_credits         0
disability              0
final_result            0
dtype: int64

# Step 6: Explore studentRegistration Dataset

In [13]:
studentRegistration.info()

<class 'pandas.DataFrame'>
RangeIndex: 32593 entries, 0 to 32592
Data columns (total 5 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   code_module          32593 non-null  str    
 1   code_presentation    32593 non-null  str    
 2   id_student           32593 non-null  int64  
 3   date_registration    32548 non-null  float64
 4   date_unregistration  10072 non-null  float64
dtypes: float64(2), int64(1), str(2)
memory usage: 1.2 MB


In [14]:
studentRegistration.head()

,code_module,code_presentation,id_student,date_registration,date_unregistration
0,AAA,2013J,11391,-159.0,NaN
1,AAA,2013J,28400,-53.0,NaN
2,AAA,2013J,30268,-92.0,12.0
3,AAA,2013J,31604,-52.0,NaN
4,AAA,2013J,32885,-176.0,NaN


In [15]:
studentRegistration.shape

(32593, 5)

In [16]:
studentRegistration.columns

Index(['code_module', 'code_presentation', 'id_student', 'date_registration',
       'date_unregistration'],
      dtype='str')

In [17]:
studentRegistration.isnull().sum()

code_module                0
code_presentation          0
id_student                 0
date_registration         45
date_unregistration    22521
dtype: int64

In [18]:
studentRegistration.duplicated().sum()

np.int64(0)

In [19]:
studentRegistration.describe()

,id_student,date_registration,date_unregistration
count,3.259300e+04,32548.000000,10072.000000
mean,7.066877e+05,-69.411300,49.757645
std,5.491673e+05,49.260522,82.460890
min,3.733000e+03,-322.000000,-365.000000
25%,5.085730e+05,-100.000000,-2.000000
50%,5.903100e+05,-57.000000,27.000000
75%,6.444530e+05,-29.000000,109.000000
max,2.716795e+06,167.000000,444.000000


# Step 7: Clean studentRegistration Dataset

## Handle Missing Values

The `studentRegistration` dataset contains missing values in two columns. The `date_registration` column has only 45 missing values, while the `date_unregistration` column has a large number of missing values. Missing values in `date_unregistration` indicate that the student did not unregister from the course, so they carry meaningful information and should not be removed.

In [20]:
studentRegistration.isnull().sum()

code_module                0
code_presentation          0
id_student                 0
date_registration         45
date_unregistration    22521
dtype: int64

In [21]:
studentRegistration["date_registration"] = studentRegistration["date_registration"].fillna(
    studentRegistration["date_registration"].median()
)

In [22]:
studentRegistration.isnull().sum()

code_module                0
code_presentation          0
id_student                 0
date_registration          0
date_unregistration    22521
dtype: int64

## Observation

The missing values in `date_registration` were replaced using the median value because only a small number of records (45) were missing. The missing values in `date_unregistration` were retained because they indicate that a student did not withdraw from the course, making them meaningful rather than erroneous.

In [23]:
studentRegistration["is_withdrawn"] = studentRegistration["date_unregistration"].notna().astype(int)

# Step 8: Explore assessments Dataset

### Explanation

The `assessments` dataset contains information about all assessments conducted in each course module, such as assessment type, weight, and due date. This dataset will later be merged with `studentAssessment` to calculate students' assessment performance.

In [24]:
assessments.info()

<class 'pandas.DataFrame'>
RangeIndex: 206 entries, 0 to 205
Data columns (total 6 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   code_module        206 non-null    str    
 1   code_presentation  206 non-null    str    
 2   id_assessment      206 non-null    int64  
 3   assessment_type    206 non-null    str    
 4   date               195 non-null    float64
 5   weight             206 non-null    float64
dtypes: float64(2), int64(1), str(3)
memory usage: 9.8 KB


### Explanation

The `info()` function provides the structure of the dataset, including the number of records, data types, and missing values.

In [25]:
assessments.head()

,code_module,code_presentation,id_assessment,assessment_type,date,weight
0,AAA,2013J,1752,TMA,19.0,10.0
1,AAA,2013J,1753,TMA,54.0,20.0
2,AAA,2013J,1754,TMA,117.0,20.0
3,AAA,2013J,1755,TMA,166.0,20.0
4,AAA,2013J,1756,TMA,215.0,30.0


### Explanation

The `head()` function displays the first five records, helping us understand the dataset structure.

In [26]:
assessments.shape

(206, 6)

### Explanation

The `shape` attribute returns the total number of rows and columns in the dataset.

In [27]:
assessments.columns

Index(['code_module', 'code_presentation', 'id_assessment', 'assessment_type',
       'date', 'weight'],
      dtype='str')

### Explanation

This command lists all column names, which will be useful when merging datasets.

In [28]:
assessments.isnull().sum()

code_module           0
code_presentation     0
id_assessment         0
assessment_type       0
date                 11
weight                0
dtype: int64

### Explanation

This command checks for missing values in each column to identify any incomplete data.

In [29]:
assessments.duplicated().sum()

np.int64(0)

### Explanation

This function checks whether duplicate records exist in the dataset.

In [30]:
assessments.describe()

,id_assessment,date,weight
count,206.000000,195.000000,206.000000
mean,26473.975728,145.005128,20.873786
std,10098.625521,76.001119,30.384224
min,1752.000000,12.000000,0.000000
25%,15023.250000,71.000000,0.000000
50%,25364.500000,152.000000,12.500000
75%,34891.750000,222.000000,24.250000
max,40088.000000,261.000000,100.000000


### Explanation

The `describe()` function provides summary statistics for numerical columns, including count, mean, minimum, maximum, and standard deviation.

## Observation

The `assessments` dataset was explored to understand its structure and quality. Missing values and duplicate records were checked before merging it with the `studentAssessment` dataset.

# Step 9: Explore studentAssessment Dataset

### Explanation

The `studentAssessment` dataset contains information about students' assessment attempts, including the assessment ID, submission date, whether the assessment was submitted, and the obtained score. This dataset is essential because it directly reflects students' academic performance and will later be aggregated to create features for machine learning models.

In [31]:
studentAssessment.info()

<class 'pandas.DataFrame'>
RangeIndex: 173912 entries, 0 to 173911
Data columns (total 5 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   id_assessment   173912 non-null  int64  
 1   id_student      173912 non-null  int64  
 2   date_submitted  173912 non-null  int64  
 3   is_banked       173912 non-null  int64  
 4   score           173739 non-null  float64
dtypes: float64(1), int64(4)
memory usage: 6.6 MB


### Explanation

The `info()` function provides the total number of records, column names, data types, and missing values in the dataset.

In [32]:
studentAssessment.head()

,id_assessment,id_student,date_submitted,is_banked,score
0,1752,11391,18,0,78.0
1,1752,28400,22,0,70.0
2,1752,31604,17,0,72.0
3,1752,32885,26,0,69.0
4,1752,38053,19,0,79.0


### Explanation

The `head()` function displays the first five records to understand the structure of the dataset.

In [33]:
studentAssessment.shape

(173912, 5)

### Explanation

The `shape` attribute returns the total number of rows and columns in the dataset.

In [34]:
studentAssessment.columns

Index(['id_assessment', 'id_student', 'date_submitted', 'is_banked', 'score'], dtype='str')

### Explanation

This command displays the names of all columns, which will later be used during dataset aggregation and merging.

In [35]:
studentAssessment.isnull().sum()

id_assessment       0
id_student          0
date_submitted      0
is_banked           0
score             173
dtype: int64

### Explanation

This function counts the missing values in each column. Identifying missing values is an important preprocessing step before aggregation.

In [36]:
studentAssessment.duplicated().sum()

np.int64(0)

### Explanation

This command checks whether duplicate records exist in the dataset.

In [37]:
studentAssessment.describe()

,id_assessment,id_student,date_submitted,is_banked,score
count,173912.000000,1.739120e+05,173912.000000,173912.000000,173739.000000
mean,26553.803556,7.051507e+05,116.032942,0.010977,75.799573
std,8829.784254,5.523952e+05,71.484148,0.104194,18.798107
min,1752.000000,6.516000e+03,-11.000000,0.000000,0.000000
25%,15022.000000,5.044290e+05,51.000000,0.000000,65.000000
50%,25359.000000,5.852080e+05,116.000000,0.000000,80.000000
75%,34883.000000,6.344980e+05,173.000000,0.000000,90.000000
max,37443.000000,2.698588e+06,608.000000,1.000000,100.000000


### Explanation

The `describe()` function provides summary statistics such as count, mean, minimum, maximum, and standard deviation for numerical columns.

## Observation

The `studentAssessment` dataset was explored to examine its structure, identify missing values, and detect duplicate records. This information will be used in the next step to aggregate assessment scores and create student-level features for machine learning.

# Step 10: Clean and Aggregate studentAssessment Dataset

### Explanation

The `studentAssessment` dataset contains multiple assessment records for each student. Before aggregation, missing values in the `score` column are removed because the score is the most important feature for measuring academic performance.

In [38]:
studentAssessment = studentAssessment.dropna(subset=["score"])

### Explanation

The missing values in the `score` column were removed because only 173 records were incomplete, representing less than 1% of the dataset.

In [39]:
studentAssessment.isnull().sum()

id_assessment     0
id_student        0
date_submitted    0
is_banked         0
score             0
dtype: int64

### Explanation

The `studentAssessment` dataset is merged with the `assessments` dataset using the `id_assessment` column. This adds assessment information such as module, presentation, assessment type, and weight to each student's assessment record.

In [40]:
studentAssessmentMerged = pd.merge(
    studentAssessment,
    assessments,
    on="id_assessment",
    how="left"
)

In [41]:
studentAssessmentMerged.head()

,id_assessment,id_student,date_submitted,is_banked,score,code_module,code_presentation,assessment_type,date,weight
0,1752,11391,18,0,78.0,AAA,2013J,TMA,19.0,10.0
1,1752,28400,22,0,70.0,AAA,2013J,TMA,19.0,10.0
2,1752,31604,17,0,72.0,AAA,2013J,TMA,19.0,10.0
3,1752,32885,26,0,69.0,AAA,2013J,TMA,19.0,10.0
4,1752,38053,19,0,79.0,AAA,2013J,TMA,19.0,10.0


### Explanation

The assessment records are grouped by student, module, and presentation to create one record per student. Features such as average score, maximum score, minimum score, and total assessments are calculated.

In [42]:
assessment_summary = studentAssessmentMerged.groupby(
    ["id_student", "code_module", "code_presentation"]
).agg(
    average_score=("score", "mean"),
    max_score=("score", "max"),
    min_score=("score", "min"),
    total_assessments=("score", "count")
).reset_index()

In [43]:
assessment_summary.head()

,id_student,code_module,code_presentation,average_score,max_score,min_score,total_assessments
0,6516,AAA,2014J,61.800000,77.0,48.0,5
1,8462,DDD,2013J,87.666667,93.0,83.0,3
2,8462,DDD,2014J,86.500000,93.0,83.0,4
3,11391,AAA,2013J,82.000000,85.0,78.0,5
4,23629,BBB,2013B,82.500000,100.0,63.0,4


In [44]:
assessment_summary.shape

(25820, 7)

## Observation

The assessment records were successfully aggregated into student-level features. Each row now represents one student enrolled in a specific module and presentation. These aggregated features will be merged with the main dataset in later steps.

## Observation

The assessment records were successfully aggregated into student-level features. The original dataset contained 173,912 assessment records, which were reduced to 25,820 student-module records after aggregation. The generated features include average score, maximum score, minimum score, and total number of assessments. These features will be used for machine learning model training.

# Step 11: Aggregate studentVle Dataset

### Explanation

The `studentVle` dataset records every interaction of a student with the Virtual Learning Environment (VLE). Since the dataset contains more than 10 million records, it is aggregated to generate student-level engagement features before merging.

In [45]:
studentVle.info()

<class 'pandas.DataFrame'>
RangeIndex: 10655280 entries, 0 to 10655279
Data columns (total 6 columns):
 #   Column             Dtype
---  ------             -----
 0   code_module        str  
 1   code_presentation  str  
 2   id_student         int64
 3   id_site            int64
 4   date               int64
 5   sum_click          int64
dtypes: int64(4), str(2)
memory usage: 487.8 MB


The `info()` function provides an overview of the dataset, including the number of records, column names, data types, and missing values.

In [46]:
studentVle.head()

,code_module,code_presentation,id_student,id_site,date,sum_click
0,AAA,2013J,28400,546652,-10,4
1,AAA,2013J,28400,546652,-10,1
2,AAA,2013J,28400,546652,-10,1
3,AAA,2013J,28400,546614,-10,11
4,AAA,2013J,28400,546714,-10,1


In [47]:
studentVle.shape

(10655280, 6)

In [48]:
studentVle.isnull().sum()

code_module          0
code_presentation    0
id_student           0
id_site              0
date                 0
sum_click            0
dtype: int64

In [49]:
studentVle.duplicated().sum()

np.int64(787170)

In [50]:
studentVle.describe()

,id_student,id_site,date,sum_click
count,1.065528e+07,1.065528e+07,1.065528e+07,1.065528e+07
mean,7.333336e+05,7.383234e+05,9.517400e+01,3.716946e+00
std,5.827060e+05,1.312196e+05,7.607130e+01,8.849047e+00
min,6.516000e+03,5.267210e+05,-2.500000e+01,1.000000e+00
25%,5.077430e+05,6.735190e+05,2.500000e+01,1.000000e+00
50%,5.882360e+05,7.300690e+05,8.600000e+01,2.000000e+00
75%,6.464840e+05,8.770300e+05,1.560000e+02,3.000000e+00
max,2.698588e+06,1.049562e+06,2.690000e+02,6.977000e+03


## Observation

The `studentVle` dataset contains students' interaction logs with the virtual learning environment. It is the largest dataset in OULAD and will be aggregated to generate engagement-based features before merging.

# Step 12: Aggregate studentVle Dataset

### Explanation

The `studentVle` dataset contains more than 10 million interaction records. Instead of merging the raw data, it is aggregated to generate student-level engagement features. This reduces the dataset size and makes it suitable for machine learning.

In [51]:
vle_summary = studentVle.groupby(
    ["id_student", "code_module", "code_presentation"]
).agg(
    total_clicks=("sum_click", "sum"),
    average_clicks=("sum_click", "mean"),
    max_clicks=("sum_click", "max"),
    total_visits=("sum_click", "count")
).reset_index()

The interaction records are grouped by student, module, and presentation. Four engagement features are created:
- **total_clicks:** Total clicks made by the student.
- **average_clicks:** Average clicks per interaction.
- **max_clicks:** Maximum clicks recorded in a single interaction.
- **total_visits:** Total number of interaction records.

In [52]:
vle_summary.head()

,id_student,code_module,code_presentation,total_clicks,average_clicks,max_clicks,total_visits
0,6516,AAA,2014J,2791,4.216012,49,662
1,8462,DDD,2013J,646,2.153333,16,300
2,8462,DDD,2014J,10,2.500000,7,4
3,11391,AAA,2013J,934,4.765306,76,196
4,23629,BBB,2013B,161,2.728814,13,59


In [53]:
vle_summary.shape

(29228, 7)

## Observation

The `studentVle` dataset was successfully aggregated into student-level engagement features. The resulting dataset contains one record for each student-module-presentation combination and is ready to be merged with the assessment and student information datasets.

# Step 12A: Explore VLE Dataset

### Explanation

The `vle` dataset contains information about learning resources available in the Virtual Learning Environment (VLE), such as the activity type and the weeks during which each resource was available. This dataset will be merged with `studentVle` to enrich student interaction data.

In [54]:
vle.info()

<class 'pandas.DataFrame'>
RangeIndex: 6364 entries, 0 to 6363
Data columns (total 6 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   id_site            6364 non-null   int64  
 1   code_module        6364 non-null   str    
 2   code_presentation  6364 non-null   str    
 3   activity_type      6364 non-null   str    
 4   week_from          1121 non-null   float64
 5   week_to            1121 non-null   float64
dtypes: float64(2), int64(1), str(3)
memory usage: 298.4 KB


In [55]:
vle.head()

,id_site,code_module,code_presentation,activity_type,week_from,week_to
0,546943,AAA,2013J,resource,NaN,NaN
1,546712,AAA,2013J,oucontent,NaN,NaN
2,546998,AAA,2013J,resource,NaN,NaN
3,546888,AAA,2013J,url,NaN,NaN
4,547035,AAA,2013J,resource,NaN,NaN


In [56]:
vle.shape

(6364, 6)

In [57]:
vle.columns

Index(['id_site', 'code_module', 'code_presentation', 'activity_type',
       'week_from', 'week_to'],
      dtype='str')

In [59]:
vle.isnull().sum()

id_site                 0
code_module             0
code_presentation       0
activity_type           0
week_from            5243
week_to              5243
dtype: int64

In [58]:
vle.duplicated().sum()

np.int64(0)

In [60]:
vle.describe()

,id_site,week_from,week_to
count,6.364000e+03,1121.000000,1121.000000
mean,7.260991e+05,15.204282,15.214987
std,1.283151e+05,8.792865,8.779806
min,5.267210e+05,0.000000,0.000000
25%,6.615928e+05,8.000000,8.000000
50%,7.300965e+05,15.000000,15.000000
75%,8.140162e+05,22.000000,22.000000
max,1.077905e+06,29.000000,29.000000


# Step 12B: Merge studentVle with VLE Dataset

### Explanation

The `studentVle` dataset contains student interaction logs, while the `vle` dataset contains information about each learning resource. These datasets are merged using `id_site`, `code_module`, and `code_presentation` to associate each student interaction with its corresponding activity type.

In [61]:
studentVleMerged = pd.merge(
    studentVle,
    vle,
    on=["id_site", "code_module", "code_presentation"],
    how="left"
)

In [62]:
studentVleMerged.head()

,code_module,code_presentation,id_student,id_site,date,sum_click,activity_type,week_from,week_to
0,AAA,2013J,28400,546652,-10,4,forumng,NaN,NaN
1,AAA,2013J,28400,546652,-10,1,forumng,NaN,NaN
2,AAA,2013J,28400,546652,-10,1,forumng,NaN,NaN
3,AAA,2013J,28400,546614,-10,11,homepage,NaN,NaN
4,AAA,2013J,28400,546714,-10,1,oucontent,NaN,NaN


In [65]:
studentVleMerged.shape

(10655280, 9)

In [63]:
studentVleMerged.isnull().sum()

code_module                0
code_presentation          0
id_student                 0
id_site                    0
date                       0
sum_click                  0
activity_type              0
week_from            9269544
week_to              9269544
dtype: int64

In [64]:
studentVleMerged["activity_type"].value_counts()

activity_type
forumng           2408457
oucontent         1963782
subpage           1949898
homepage          1735226
quiz               914573
resource           813185
url                407716
ouwiki             178165
oucollaborate       65793
externalquiz        50910
page                47549
questionnaire       44905
ouelluminate        22360
glossary            17435
dataplus            13506
dualpane            10153
htmlactivity         6835
folder               4678
sharedsubpage         148
repeatactivity          6
Name: count, dtype: int64

# Step 12C: Generate Student Engagement Features

The merged VLE dataset contains different types of learning activities such as forums, quizzes, homepage visits, and learning resources. These interactions are aggregated to generate engagement-based features for each student.

In [66]:
vle_summary = studentVleMerged.groupby(
    ["id_student", "code_module", "code_presentation"]
).agg(
    total_clicks=("sum_click", "sum"),
    average_clicks=("sum_click", "mean"),
    max_clicks=("sum_click", "max"),
    total_visits=("sum_click", "count")
).reset_index()

In [67]:
activity_summary = studentVleMerged.pivot_table(
    index=["id_student", "code_module", "code_presentation"],
    columns="activity_type",
    values="sum_click",
    aggfunc="sum",
    fill_value=0
).reset_index()

In [68]:
activity_summary.head()

activity_type,id_student,code_module,code_presentation,dataplus,dualpane,externalquiz,folder,forumng,glossary,homepage,...,ouelluminate,ouwiki,page,questionnaire,quiz,repeatactivity,resource,sharedsubpage,subpage,url
0,6516,AAA,2014J,21,0,0,0,451,0,497,...,0,0,0,0,0,0,31,0,143,143
1,8462,DDD,2013J,0,0,12,0,36,0,184,...,0,18,0,0,0,0,70,0,227,23
2,8462,DDD,2014J,0,0,0,0,2,0,7,...,0,0,0,0,0,0,0,0,0,0
3,11391,AAA,2013J,0,0,0,0,193,0,138,...,0,0,0,0,0,0,13,0,32,5
4,23629,BBB,2013B,0,0,0,0,87,0,36,...,0,0,0,0,31,0,2,0,5,0


In [69]:
activity_summary.shape

(29228, 23)

# Step 13: Merge All OULAD Datasets

### Explanation

The studentInfo dataset is used as the base dataset because it contains one record for each student enrolled in a module. The remaining datasets are merged using left joins to preserve all students while enriching the dataset with registration, assessment, engagement, activity, and course information.

In [70]:
final_dataset = pd.merge(
    studentInfo,
    studentRegistration,
    on=["id_student", "code_module", "code_presentation"],
    how="left"
)

In [71]:
final_dataset = pd.merge(
    final_dataset,
    assessment_summary,
    on=["id_student", "code_module", "code_presentation"],
    how="left"
)

In [72]:
final_dataset = pd.merge(
    final_dataset,
    vle_summary,
    on=["id_student", "code_module", "code_presentation"],
    how="left"
)

In [73]:
final_dataset = pd.merge(
    final_dataset,
    activity_summary,
    on=["id_student", "code_module", "code_presentation"],
    how="left"
)

In [74]:
final_dataset = pd.merge(
    final_dataset,
    courses,
    on=["code_module", "code_presentation"],
    how="left"
)

In [75]:
final_dataset.info()

<class 'pandas.DataFrame'>
RangeIndex: 32593 entries, 0 to 32592
Data columns (total 44 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   code_module                 32593 non-null  str    
 1   code_presentation           32593 non-null  str    
 2   id_student                  32593 non-null  int64  
 3   gender                      32593 non-null  str    
 4   region                      32593 non-null  str    
 5   highest_education           32593 non-null  str    
 6   imd_band                    32593 non-null  str    
 7   age_band                    32593 non-null  str    
 8   num_of_prev_attempts        32593 non-null  int64  
 9   studied_credits             32593 non-null  int64  
 10  disability                  32593 non-null  str    
 11  final_result                32593 non-null  str    
 12  date_registration           32593 non-null  float64
 13  date_unregistration         10072 non-null

In [76]:
final_dataset.head()

,code_module,code_presentation,id_student,gender,region,highest_education,imd_band,age_band,num_of_prev_attempts,studied_credits,...,ouwiki,page,questionnaire,quiz,repeatactivity,resource,sharedsubpage,subpage,url,module_presentation_length
0,AAA,2013J,11391,M,East Anglian Region,HE Qualification,90-100%,55<=,0,240,...,0.0,0.0,0.0,0.0,0.0,13.0,0.0,32.0,5.0,268
1,AAA,2013J,28400,F,Scotland,HE Qualification,20-30%,35-55,0,60,...,0.0,0.0,0.0,0.0,0.0,12.0,0.0,87.0,48.0,268
2,AAA,2013J,30268,F,North Western Region,A Level or Equivalent,30-40%,35-55,0,60,...,0.0,0.0,0.0,0.0,0.0,4.0,0.0,22.0,4.0,268
3,AAA,2013J,31604,F,South East Region,A Level or Equivalent,50-60%,35-55,0,60,...,0.0,0.0,0.0,0.0,0.0,19.0,0.0,144.0,90.0,268
4,AAA,2013J,32885,F,West Midlands Region,Lower Than A Level,50-60%,0-35,0,60,...,0.0,0.0,0.0,0.0,0.0,45.0,0.0,79.0,14.0,268


In [77]:
final_dataset.shape

(32593, 44)

In [78]:
final_dataset.isnull().sum()

code_module                       0
code_presentation                 0
id_student                        0
gender                            0
region                            0
highest_education                 0
imd_band                          0
age_band                          0
num_of_prev_attempts              0
studied_credits                   0
disability                        0
final_result                      0
date_registration                 0
date_unregistration           22521
is_withdrawn                      0
average_score                  6773
max_score                      6773
min_score                      6773
total_assessments              6773
total_clicks                   3365
average_clicks                 3365
max_clicks                     3365
total_visits                   3365
dataplus                       3365
dualpane                       3365
externalquiz                   3365
folder                         3365
forumng                     

## Observation

All seven OULAD datasets were successfully merged into a single integrated dataset. Each row represents one student enrolled in a specific module and presentation. The dataset now contains demographic information, registration details, assessment performance, engagement metrics, activity-wise interaction features, and course information, making it suitable for machine learning.

# Step 14: Handle Missing Values

The merged dataset contains missing values because some students did not attempt assessments or did not interact with the Virtual Learning Environment. These missing values are replaced with zero because the absence of activity indicates no participation rather than missing information.

In [79]:
assessment_cols = [
    "average_score",
    "max_score",
    "min_score",
    "total_assessments"
]

final_dataset[assessment_cols] = final_dataset[assessment_cols].fillna(0)

In [80]:
vle_cols = [
    "total_clicks",
    "average_clicks",
    "max_clicks",
    "total_visits",
    "dataplus",
    "dualpane",
    "externalquiz",
    "folder",
    "forumng",
    "glossary",
    "homepage",
    "htmlactivity",
    "oucollaborate",
    "oucontent",
    "ouelluminate",
    "ouwiki",
    "page",
    "questionnaire",
    "quiz",
    "repeatactivity",
    "resource",
    "sharedsubpage",
    "subpage",
    "url"
]

final_dataset[vle_cols] = final_dataset[vle_cols].fillna(0)

In [81]:
final_dataset.isnull().sum()

code_module                       0
code_presentation                 0
id_student                        0
gender                            0
region                            0
highest_education                 0
imd_band                          0
age_band                          0
num_of_prev_attempts              0
studied_credits                   0
disability                        0
final_result                      0
date_registration                 0
date_unregistration           22521
is_withdrawn                      0
average_score                     0
max_score                         0
min_score                         0
total_assessments                 0
total_clicks                      0
average_clicks                    0
max_clicks                        0
total_visits                      0
dataplus                          0
dualpane                          0
externalquiz                      0
folder                            0
forumng                     

## Observation

Missing values in assessment and VLE features were replaced with zero because they represent the absence of assessments or online learning activities rather than incomplete data. The missing values in `date_unregistration` were retained because they indicate that a student did not withdraw from the course.